elastic/kibana
# [Fleet] Add agent incoming data endpoint and presentational component

## Summary

Part of https://github.com/elastic/kibana/issues/125534

- Adding an endpoint `agent_status/data` that queries `logs-*-*,metrics-*-*,traces-*-*,synthetics-*-*` to verify that a set of agents is receiving data since the last 5 minutes.

-  Querying the new endpoint from a react component to signal that a newly enrolled agent has received data.  


Received data (two possible buttons are shown)
![Screenshot 2022-03-10 at 17 05 57](https://user-images.githubusercontent.com/16084106/157704150-f930788c-b542-4851-8e7b-26b2e265b84a.png)

![Screenshot 2022-03-10 at 17 07 03](https://user-images.githubusercontent.com/16084106/157704489-cbfec229-792d-491a-b3cb-2e20962b5535.png)

Loading:
![Screenshot 2022-03-08 at 18 20 23](https://user-images.githubusercontent.com/16084106/157291128-766bc077-b775-41de-b046-0c9aa318150c.png)

#### Testing steps

- Make sure you have an agent running and ingesting data
- Send a request like this
```
  curl --location --request GET 'localhost:5601/<YOUR_PATH>/api/fleet/agent_status/data?agentsId=[<ANEXISTINGAGENTID>, <ANOTHEREXISTINGAGENTID>]' -u elastic:changeme
```
  - You should get a response similar to this one:
```
    {"items": [
        {
        "bb59252d-ad7f-4289-9dbe-6808582638f6":
            {
               "data": true
             }
        },
        {
        "9a6edd86-ab0e-4683-8706-2a2a0e36b977":
            {
              "data": false
            }
        }
     ]
   }
```

To test the component I added it [here](https://github.com/elastic/kibana/blob/9ddb3e8756e9294c857e58b3ab0deedfb9375cbe/x-pack/plugins/fleet/public/components/agent_enrollment_flyout/index.tsx#L102) where agentsIds is an array of strings (the agents ids to check).

- Test the case when the button shows "Install APM agent":

`    <ConfirmIncomingData
          agentsIds={agentsCheckData}
          installedPolicy={{ name: 'apm', version: '1.2.0' }}
        />`

- Test the case when the button shows "View Assets":

`    <ConfirmIncomingData
          agentsIds={agentsCheckData}
          installedPolicy={{ name: 'nginx', version: '1.2.0' }}
        />`
- Test the case when no button is shown (similar to current behavior in "add agent"):
`    <ConfirmIncomingData
          agentsIds={agentsCheckData}
        />
`

#### Example of the ES query used:
```
POST /logs-*-*,metrics-*-*,traces-*-*,synthetics-*-*/_search?allow_partial_search_results=true&_source=false&timeout=5s&size=0
{
  "query": {
    "bool": {
      "filter": [
        {
          "terms": {
            "agent.id": [
              "fa892bb4-b7da-4022-8aaa-4126ba9d2362",
              "9a6edd86-ab0e-4683-8706-2a2a0e36b977"
            ]
          }
        },
        {
          "range": {
            "@timestamp": {
              "gte": "now-5m",
              "lte": "now"
            }
          }
        }
      ]
    }
  },
  "aggs": {
    "agent_ids": {
      "terms": {
        "field": "agent.id",
        "size": 5
      }
    }
  }
}
```


### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
